packages/ckeditor5-utils/src/splicearray.ts

```diff
@@ -0,0 +1,39 @@
+/**
+ * @license Copyright (c) 2003-2022, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module utils/splicearray
+ */
+
+/**
+ * Function that splices Arrays and could be used instead of Array.prototype.splice as it results
+ * into "Maximum call stack size exceeded" when dealing with very big number.
+ *
+ * @private
+ * @param {Array} target Array to be spliced.
+ * @param {Array} source Array of elements to be inserted to target.
+ * @param {Number} start Index at which nodes should be inserted/removed.
+ * @param {Number} count Number of items.
+ * @param {Number} [chunkSize=10000] Chunk size of source array.
+ *
+ * 		spliceArray( [ 1, 2 ], [ 3, 4 ], 0, 0 );		// [ 3, 4, 1, 2]
+ * 		spliceArray( [ 1, 2 ], [ 3, 4 ], 1, 1 );		// [ 1, 3, 4 ]
+ * 		spliceArray( [ 1, 2 ], [ 3, 4 ], 1, 0 );		// [ 1, 3, 4, 2 ]
+ * 		spliceArray( [ 1, 2 ], [ 3, 4 ], 2, 0 );		// [ 1, 2, 3, 4 ]
+ * 		spliceArray( [ 1, 2 ], [], 0, 1 );				// [ 2 ]
+ *
+ */
+export default function spliceArray<T>( target: Array<T>, source: Array<T>, start: number, count: number, chunkSize = 10000 ): Array<T> {
```

Comment:
I have doubts about `chunkSize`. Is it really needed? Do you see a scenario where you might want to use different chunk sizes? It could be simply hardcoded in a property in this file.

Also, if we would change the implementation, for example to:

```js
return target.slice( ... ).concat( source ).concat( target.slice( ... ) );
```

The parameter would be useless and would pollute code. To me, the parameter looks like implementation detail and not something you would want to change in various calls.

BTW. I know you did some research on this problem. I wonder if you found any solution using slice+concat or maybe you tried your own perf tests? I am asking because slice+concat solution looks much more elegant to me when I look at it now (I know I suggested the implemented solution in the beginning :)).

Reply:
I'm back with some info. I implemented the new method as you proposed (let's call it `proposal`) and then compared it with my method (let's call it `actual`) and the conclusions are interesting. I used jsbench to compare the performance providing different datasets:

1.  `target.length = 2`    `source.length = 2`  
    Result: `proposal` is **27,48% slower** 
2.  `target.length = 2`    `source.length = 10000`  
    Result: `proposal` is **66,80% faster**
3.  `target.length = 10`    `source.length = 2`  
    Result: `proposal` is **22,74% slower** 
4.  `target.length = 10 000`  `source.length = 2`  
    Result: `proposal` is **42,43% faster**
5.  `target.length = 10 000`  `source.length = 10000`  
    Result: `proposal` is **58,33% faster**

As you can see... the more data we process the slower `actual` solution works.  
  
**Proposal:** we can implement some kind of hybrid method which will perform normal `Array.prototype.splice` for small set of data and the `propsal` solution for large arrays (because of `splice` method).  
**Drawback:** `Array.prototype.splice` works with reference and `proposal` solution returns new Array.